src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -297,13 +297,13 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
                         if (_state != StreamState.ExpectingHeaders && _state != StreamState.ExpectingTrailingHeaders)
                         {
                             if (NetEventSource.IsEnabled) Trace("Received header before status.");
-                            throw new Http2ProtocolException(SR.net_http_invalid_response);
+                            throw new HttpRequestException(SR.net_http_invalid_response);
```

Comment:
I'm not clear on when HttpRequestException is used directly vs an Http2ConnectionException (which will then end up getting wrapped in an HttpRequestException higher up).  What makes this case different, for example, from the `if (_state != StreamState.ExpectingStatus && _state != StreamState.ExpectingData)` case below that uses Http2ConnectionException?

Reply:
It's a little vague at the moment.

The intention is to use Http2ProtocolException for protocol violations that are connection-level errors and unrecoverable, because they effectively put the connection in a bad, unusable state. So for example, in this case, we've received a HEADERS block where the status code is not the first entry. This is an invalid HEADERS block, and we need to fail the request. But it's not an HTTP2 protocol violation per se; it doesn't need to kill the connection. Another example is an invalid header name. In the case below, we've received a HEADERS block at a point where it is not allowed per HTTP2 spec and thus the connection is in a bad state and must be killed.

Another way to look at it is, Http2ConnectionException represents protocol violations, whereas HttpRequestException represents semantic violations.

All that said, this is complicated by a couple things:
(1) Even when we throw HttpRequestException, we're going to kill the connection instead of just killing the request. This isn't ideal, but the actual errors here are fairly obscure so I think it's not a big deal.
(2) This is all complicated somewhat by the fact that the _state on Http2Stream currently represents both the "stream state" as per HTTP2 spec as well as some internal states we use for tracking header processing, e.g. ExpectingStatus. I think it would be better to clean some of this up, but it's not urgent.